### PR TITLE
ensure that all discharge macaroons are used exactly once

### PR DESCRIPTION
**Pull request by rogpeppe**
This means means that if we get a successful discharge, we can be sure
that all the first party caveats presented in the macaroon slice have
been checked.

### CODE REVIEW

**Comment by mhilton** [comment]
:+1:

File: macaroon_test.go [hunk 1]
```diff
@@ -324,6 +327,18 @@ var verifyTests = []struct {
...
+	macaroons: []macaroonSpec{{
+		rootKey: "root-key",
+		id:      "root-id",
+	}, {
+		rootKey: "other-key",
+		id:      "unused",
+	}},
+	conditions: []conditionTest{{
+		expectErr: `discharge macaroon "unused" was not used`,
+	}},
```
Comment (frankban) [inline macaroon_test.go:395]: Is it possible to test the case in which a discharge macaroon is used more than once?

**Comment by frankban** [comment]
:+1: with a question.
Very nice, thank you!

File: macaroon.go [hunk 2]
```diff
@@ -232,24 +242,26 @@ func (m *Macaroon) verify(rootSig []byte, rootKey []byte, check func(caveat stri
...
 					continue
 				}
 				found = true
-				verifyErr = dm.verify(rootSig, cavKey, check, discharges)
-				if verifyErr == nil {
-					break
+
+				// It's important that we do this before calling verify,
+				// as it prevents potentially infinite recursion.
+				if used[di]++; used[di] > 1 {
```
Comment (frankban) [inline macaroon.go:254]: Nice.

**Comment by frankban** [comment]
:+1: thanks for fixing the infinite recursion.

**Pull request merged**